File: clients/C++/5.2 mcp_delete_user/MCPDeleteUserClient.cpp
#include <iostream>
#include <string>
#include <map>
#include <sstream>
#include <stdexcept>
#include <json/json.h>
#include <winsock2.h>
#include <ws2tcpip.h>

#pragma comment(lib, "ws2_32.lib") // Verlinkung mit der Winsock-Bibliothek

// Funktion zum Parsen von Argumenten
std::map<std::string, std::string> parseArguments(int argc, char* argv[]) {
    std::map<std::string, std::string> args;
    for (int i = 1; i < argc; ++i) {
        std::string key = argv[i];
        if (i + 1 < argc && key.rfind("--", 0) == 0) {
            args[key] = argv[++i];
        }
    }
    return args;
}

// Funktion zum Senden der Anfrage
std::string sendRequest(const std::string& serverIp, int serverPort, const Json::Value& payload) {
    Json::StreamWriterBuilder writer;
    std::string payloadJson = Json::writeString(writer, payload);

    // Winsock initialisieren
    WSADATA wsaData;
    if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
        throw std::runtime_error("Failed to initialize Winsock.");
    }

    // Socket erstellen
    SOCKET sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock == INVALID_SOCKET) {
        WSACleanup();
        throw std::runtime_error("Failed to create socket.");
    }

    // Server-Adresse konfigurieren
    sockaddr_in serverAddr;
    serverAddr.sin_family = AF_INET;
    serverAddr.sin_port = htons(serverPort);
    if (inet_pton(AF_INET, serverIp.c_str(), &serverAddr.sin_addr) <= 0) {
        closesocket(sock);
        WSACleanup();
        throw std::runtime_error("Invalid server IP address.");
    }

    // Verbindung herstellen
    if (connect(sock, (sockaddr*)&serverAddr, sizeof(serverAddr)) < 0) {
        closesocket(sock);
        WSACleanup();
        throw std::runtime_error("Connection failed.");
    }

    // Daten senden
    if (send(sock, payloadJson.c_str(), payloadJson.size(), 0) < 0) {
        closesocket(sock);
        WSACleanup();
        throw std::runtime_error("Failed to send data.");
    }

    // Antwort empfangen
    char buffer[4096];
    int bytesRead;
    std::ostringstream response;

    do {
        bytesRead = recv(sock, buffer, sizeof(buffer) - 1, 0);
        if (bytesRead > 0) {
            buffer[bytesRead] = '\0'; // Null-terminieren
            response << buffer;
        }
    } while (bytesRead == sizeof(buffer) - 1);

    // Socket schließen
    closesocket(sock);
    WSACleanup();

    return response.str();
}

int main(int argc, char* argv[]) {
    try {
        auto args = parseArguments(argc, argv);

        // Pflichtargumente überprüfen
        if (args["--server-ip"].empty() || args["--server-port"].empty() || args["--email"].empty() || args["--token"].empty()) {
            std::cerr << "Usage: --server-ip <IP> --server-port <PORT> --email <EMAIL> --token <TOKEN>\n";
            return 1;
        }

        // Argumente auslesen
        std::string serverIp = args["--server-ip"];
        int serverPort = std::stoi(args["--server-port"]);
        std::string email = args["--email"];
        std::string token = args["--token"];

        // JSON-Payload erstellen
        Json::Value payload;
        payload["command"] = "delete_user";
        payload["token"] = token;
        payload["arguments"]["email"] = email;

        std::cout << "📤 Sending delete user request...\n";

        // Anfrage senden und Antwort erhalten
        std::string response = sendRequest(serverIp, serverPort, payload);

        std::cout << "✔️ Response from server:\n" << response << "\n";
    } catch (const std::exception& e) {
        std::cerr << "❌ ERROR: " << e.what() << "\n";
        return 1;
    }

    return 0;
}

File: clients/Java/5.2 mcp_delete_user/MCPDeleteUserClient.java
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

public class MCPDeleteUserClient {
    public static void main(String[] args) {
        // Wir benötigen 4 Schlüssel (jeweils 2 Argumente): "--server-ip <IP>", 
        // "--server-port <PORT>", "--email <EMAIL>", "--token <TOKEN>".
        // => mind. 8 Elemente in args
        if (args.length < 8) {
            System.out.println("Usage: --server-ip <IP> --server-port <PORT> --email <EMAIL> --token <TOKEN>");
            return;
        }

        // Argumente auslesen
        String serverIp = getArgument(args, "--server-ip");
        String portStr  = getArgument(args, "--server-port");
        String email    = getArgument(args, "--email");
        String token    = getArgument(args, "--token");

        // Fehlende Pflichtargumente abfangen
        if (serverIp == null || portStr == null || email == null || token == null) {
            System.err.println("Fehler: Mindestens eines der erforderlichen Argumente ist nicht gesetzt.");
            return;
        }

        int serverPort;
        try {
            serverPort = Integer.parseInt(portStr);
        } catch (NumberFormatException e) {
            System.err.println("Fehler: Server-Port muss eine ganze Zahl sein.");
            return;
        }

        // JSON-Payload vorbereiten
        JSONObject arguments = new JSONObject();
        arguments.put("email", email);

        JSONObject payload = new JSONObject();
        payload.put("command", "delete_user");
        payload.put("token", token);
        payload.put("arguments", arguments);

        System.out.println("📤 Sending delete user request...");
        String response = sendRequest(serverIp, serverPort, payload.toString());
        System.out.println("✔️ Response from server:");
        System.out.println(response);
    }

    /**
     * Holt den Wert eines bestimmten Arguments aus args, z.B. --server-ip 127.0.0.1
     * @param args Array mit allen Argumenten
     * @param key  Name des gesuchten Arguments, z.B. "--server-ip"
     * @return     Der direkt folgende Wert oder null, wenn nicht vorhanden
     */
    private static String getArgument(String[] args, String key) {
        for (int i = 0; i < args.length; i++) {
            if (key.equals(args[i]) && i < args.length - 1) {
                return args[i + 1];
            }
        }
        return null;
    }

    /**
     * Stellt eine Socket-Verbindung her, sendet den JSON-String und empfängt die Server-Antwort.
     *
     * @param serverIp   IP des Servers
     * @param serverPort Port des Servers
     * @param payload    Zu sendender JSON-String
     * @return           Server-Antwort oder Fehlermeldung
     */
    private static String sendRequest(String serverIp, int serverPort, String payload) {
        try (Socket socket = new Socket(serverIp, serverPort)) {
            // JSON an den Server senden
            OutputStream output = socket.getOutputStream();
            byte[] data = payload.getBytes(StandardCharsets.UTF_8);
            output.write(data);
            output.flush();

            // Antwort vom Server lesen
            BufferedReader reader = new BufferedReader(
                new InputStreamReader(socket.getInputStream(), StandardCharsets.UTF_8)
            );
            StringBuilder sb = new StringBuilder();
            char[] buffer = new char[4096];
            int charsRead;
            do {
                charsRead = reader.read(buffer);
                if (charsRead > 0) {
                    sb.append(buffer, 0, charsRead);
                }
            } while (charsRead == buffer.length);

            return sb.toString();
        } catch (Exception e) {
            return "Error: " + e.getMessage();
        }
    }
}
